file/s3file/s3file.go

```diff
@@ -1396,6 +1351,49 @@ func ParseURL(url string) (scheme, bucket, key string, err error) {
 	return scheme, parts[0], parts[1], nil
 }
 
+func stat(ctx context.Context, clients []s3iface.S3API, policy retryPolicy, path string) (*s3Info, error) {
+	_, bucket, key, err := ParseURL(path)
+	if err != nil {
+		return nil, errors.E(errors.Invalid, "could not parse", path, err)
+	}
+	for {
+		var ids s3RequestIDs
+		output, err := policy.client().HeadObjectWithContext(ctx,
+			&s3.HeadObjectInput{
+				Bucket: aws.String(bucket),
+				Key:    aws.String(key),
+			},
+			captureRequestIDs(&ids),
+		)
+		if policy.shouldRetry(ctx, err, path) {
+			continue
+		}
+		if err != nil {
+			return nil, annotate(err, ids, &policy, "s3file.stat", path)
+		}
+		if output.ETag == nil || *output.ETag == "" {
+			return nil, errors.E("s3file.stat: empty ETag", path, errors.NotExist, "awsrequestID:", ids.String())
+		}
+		if output.ContentLength == nil {
+			return nil, errors.E("s3file.stat: nil ContentLength", path, errors.NotExist, "awsrequestID:", ids.String())
+		}
+		if *output.ContentLength == 0 && strings.HasSuffix(path, "/") {
+			// Assume this is a directory marker:
+			// https://web.archive.org/web/20190424231712/https://docs.aws.amazon.com/AmazonS3/latest/user-guide/using-folders.html
+			return nil, errors.E("s3file.stat: directory marker at path", path, errors.Invalid, "awsrequestID:", ids.String())
```

Comment:
It looks like this code used to be `errors.NotExist`?

Reply:
This is true, but it seemed like the wrong classification to me. I thought the risk of this actually breaking someone was very low, but it is backwards-incompatible.

Reply:
I'd be concerned about breaking code that depends on this behavior for directory crawling, etc. It seems like the sort of thing that might not be well-integration-tested in some applications. Your call, though.

Reply:
Okay; at a minimum it belongs in a separate PR.